File: python/examples/openmm_u_kln_mbar.py
"""Pure-OpenMM equilibrium MBAR example for alchemrs.

This file avoids `openmmtools` so it can run with just:

    pip install openmm

It uses the same one-particle harmonic oscillator toy system as the NES
example, but samples equilibrium configurations at several lambda windows,
evaluates the reduced potential of every configuration at every lambda state,
and converts the resulting `u_kln[k][l][n]` tensor into the `UNkMatrix`
windows expected by `alchemrs`.
"""

from __future__ import annotations

from dataclasses import dataclass
from math import sqrt
import random

import alchemrs as ar
import numpy as np

from openmm import Context, CustomExternalForce, LangevinMiddleIntegrator, Platform, System, Vec3
from openmm import unit


K_B_KCAL_PER_MOL_K = 0.00198720425864083


@dataclass
class EquilibriumConfig:
    temperature: unit.Quantity = 300 * unit.kelvin
    friction: unit.Quantity = 10.0 / unit.picoseconds
    timestep: unit.Quantity = 2.0 * unit.femtoseconds
    equilibration_steps: int = 500
    steps_per_sample: int = 20
    n_samples_per_window: int = 200
    spring_constant: unit.Quantity = (
        10.0 * unit.kilocalories_per_mole / unit.angstroms**2
    )
    mass: unit.Quantity = 39.9 * unit.amu
    x_start: unit.Quantity = 0.0 * unit.angstroms
    x_end: unit.Quantity = 5.0 * unit.angstroms
    lambdas: tuple[float, ...] = (0.0, 0.5, 1.0)
    platform_name: str | None = None


def make_harmonic_system(config: EquilibriumConfig) -> System:
    system = System()
    system.addParticle(config.mass)

    force = CustomExternalForce("0.5 * k * (x - x0)^2")
    force.addGlobalParameter(
        "k", config.spring_constant.value_in_unit(unit.kilocalories_per_mole / unit.nanometers**2)
    )
    force.addGlobalParameter("x0", config.x_start.value_in_unit(unit.nanometers))
    force.addParticle(0, [])
    system.addForce(force)

    return system


def make_context(config: EquilibriumConfig) -> Context:
    system = make_harmonic_system(config)
    integrator = LangevinMiddleIntegrator(
        config.temperature, config.friction, config.timestep
    )
    platform = (
        Platform.getPlatformByName(config.platform_name)
        if config.platform_name is not None
        else Platform.getPlatformByName("Reference")
    )
    context = Context(system, integrator, platform)

    sigma = sqrt(
        (K_B_KCAL_PER_MOL_K * config.temperature.value_in_unit(unit.kelvin))
        / config.spring_constant.value_in_unit(unit.kilocalories_per_mole / unit.angstroms**2)
    )
    dx = random.gauss(0.0, sigma) * unit.angstroms
    context.setPositions([Vec3((config.x_start + dx).value_in_unit(unit.nanometers), 0.0, 0.0)])
    context.setVelocitiesToTemperature(config.temperature)
    return context


def lambda_to_x0_nm(config: EquilibriumConfig, lam: float) -> float:
    x_start_nm = config.x_start.value_in_unit(unit.nanometers)
    x_end_nm = config.x_end.value_in_unit(unit.nanometers)
    return x_start_nm + lam * (x_end_nm - x_start_nm)


def reduced_potential(context: Context, temperature_k: float) -> float:
    state = context.getState(getEnergy=True)
    energy = state.getPotentialEnergy().value_in_unit(unit.kilocalories_per_mole)
    return energy / (K_B_KCAL_PER_MOL_K * temperature_k)


def kt_in_kcal_per_mol(temperature_k: float) -> float:
    return K_B_KCAL_PER_MOL_K * temperature_k


def collect_u_kln(config: EquilibriumConfig) -> np.ndarray:
    temperature_k = config.temperature.value_in_unit(unit.kelvin)
    n_windows = len(config.lambdas)
    u_kln = np.zeros((n_windows, n_windows, config.n_samples_per_window), dtype=float)

    for sampled_idx, sampled_lambda in enumerate(config.lambdas):
        context = make_context(config)
        integrator = context.getIntegrator()
        context.setParameter("x0", lambda_to_x0_nm(config, sampled_lambda))
        integrator.step(config.equilibration_steps)

        for sample_idx in range(config.n_samples_per_window):
            integrator.step(config.steps_per_sample)
            for evaluated_idx, evaluated_lambda in enumerate(config.lambdas):
                context.setParameter("x0", lambda_to_x0_nm(config, evaluated_lambda))
                u_kln[sampled_idx, evaluated_idx, sample_idx] = reduced_potential(
                    context, temperature_k
                )
            context.setParameter("x0", lambda_to_x0_nm(config, sampled_lambda))

    return u_kln


def windows_from_u_kln(
    u_kln: np.ndarray,
    lambdas: tuple[float, ...],
    temperature_k: float,
) -> list[ar.UNkMatrix]:
    if u_kln.ndim != 3:
        raise ValueError("u_kln must have shape (sampled_state, evaluated_state, sample)")
    if u_kln.shape[0] != len(lambdas) or u_kln.shape[1] != len(lambdas):
        raise ValueError("u_kln state dimensions must match the lambda grid")

    evaluated_states = [
        ar.StatePoint([lam], temperature_k)
        for lam in lambdas
    ]

    windows: list[ar.UNkMatrix] = []
    for sampled_idx, sampled_lambda in enumerate(lambdas):
        data = u_kln[sampled_idx].T.copy()
        time_ps = np.arange(data.shape[0], dtype=float)
        windows.append(
            ar.UNkMatrix(
                data=data,
                time_ps=time_ps,
                sampled_state=ar.StatePoint([sampled_lambda], temperature_k),
                evaluated_states=evaluated_states,
            )
        )

    return windows


def main() -> None:
    random.seed(0)
    np.random.seed(0)

    config = EquilibriumConfig()
    temperature_k = config.temperature.value_in_unit(unit.kelvin)
    conversion = kt_in_kcal_per_mol(temperature_k)

    u_kln = collect_u_kln(config)
    windows = windows_from_u_kln(u_kln, config.lambdas, temperature_k)

    fit = ar.MBAR().fit(windows)
    result = fit.result_with_uncertainty()

    print("OpenMM u_kln -> alchemrs MBAR")
    print(
        "  dG: "
        f"{result.values[0, -1] * conversion:.6f} +/- "
        f"{result.uncertainties[0, -1] * conversion:.6f} kcal/mol"
    )
    print(f"  overlap scalar: {fit.overlap_scalar():.6f}")
    print(f"  windows: {len(windows)}")
    print(f"  samples per window: {config.n_samples_per_window}")


if __name__ == "__main__":
    main()

File: examples/openmm_u_kln_mbar.rs
use alchemrs::{MbarEstimator, MbarOptions, StatePoint, UNkMatrix};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // OpenMM workflows commonly accumulate reduced potentials as u_kln[k][l][n]:
    // sampled state k, evaluated state l, sample n.
    //
    // This example shows how to convert that tensor into the per-window UNkMatrix
    // objects expected by alchemrs so the existing MBAR APIs can be used directly.
    let lambdas = vec![1.0, 0.5, 0.0];
    let temperature_k = 120.0;
    let u_kln = vec![
        vec![
            vec![0.0, 0.1, 0.2, 0.1],
            vec![0.6, 0.7, 0.8, 0.7],
            vec![1.5, 1.7, 1.8, 1.6],
        ],
        vec![
            vec![0.5, 0.6, 0.7, 0.6],
            vec![0.0, 0.1, 0.1, 0.0],
            vec![0.6, 0.7, 0.8, 0.7],
        ],
        vec![
            vec![1.3, 1.4, 1.5, 1.4],
            vec![0.5, 0.6, 0.7, 0.6],
            vec![0.0, 0.1, 0.2, 0.1],
        ],
    ];

    let windows = windows_from_openmm_u_kln(&u_kln, &lambdas, temperature_k)?;
    let fit = MbarEstimator::new(MbarOptions::default()).fit(&windows)?;
    let result = fit.result_with_uncertainty()?;
    let delta_index = result.n_states() - 1;

    println!(
        "MBAR dG = {:.6} +/- {:.6} kT",
        result.values()[delta_index],
        result.uncertainties().unwrap()[delta_index]
    );
    println!("overlap scalar = {:.6}", fit.overlap_scalar()?);

    Ok(())
}

fn windows_from_openmm_u_kln(
    u_kln: &[Vec<Vec<f64>>],
    lambdas: &[f64],
    temperature_k: f64,
) -> Result<Vec<UNkMatrix>, Box<dyn std::error::Error>> {
    if u_kln.is_empty() {
        return Err("u_kln must contain at least one sampled state".into());
    }
    if u_kln.len() != lambdas.len() {
        return Err("u_kln sampled-state dimension must match lambdas".into());
    }

    let evaluated_states = lambdas
        .iter()
        .map(|&lambda| StatePoint::new(vec![lambda], temperature_k))
        .collect::<alchemrs::Result<Vec<_>>>()?;
    let n_states = evaluated_states.len();

    let mut windows = Vec::with_capacity(u_kln.len());
    for (sampled_idx, evaluated_by_state) in u_kln.iter().enumerate() {
        if evaluated_by_state.len() != n_states {
            return Err(format!(
                "u_kln[{sampled_idx}] has {} evaluated states, expected {n_states}",
                evaluated_by_state.len()
            )
            .into());
        }

        let n_samples = evaluated_by_state[0].len();
        if n_samples == 0 {
            return Err(format!("u_kln[{sampled_idx}] has no samples").into());
        }
        if evaluated_by_state
            .iter()
            .any(|series| series.len() != n_samples)
        {
            return Err(format!("u_kln[{sampled_idx}] is ragged across evaluated states").into());
        }

        let mut sample_major = Vec::with_capacity(n_samples * n_states);
        for sample_idx in 0..n_samples {
            for evaluated_idx in 0..n_states {
                sample_major.push(evaluated_by_state[evaluated_idx][sample_idx]);
            }
        }

        let time_ps = (0..n_samples).map(|idx| idx as f64).collect::<Vec<_>>();
        let sampled_state = evaluated_states[sampled_idx].clone();
        windows.push(UNkMatrix::new(
            n_samples,
            n_states,
            sample_major,
            time_ps,
            Some(sampled_state),
            evaluated_states.clone(),
        )?);
    }

    Ok(windows)
}
